### ISSUE TITLE
add support for parallel diversity calculations

### ISSUE THREAD

**Issue by gregcaporaso** [biocore/scikit-bio#1181]
This won't be addressed in #1040 as there are bigger questions about how we generally want to support parallelization in scikit-bio. 

We'll ultimately want to be support computing alpha diversity for multiple samples and computing beta diversity for multiple pairs of samples in parallel.

@wasade proposed an interface for this before I started working on that branch, but I can't find it now. @wasade, did you squash commits, and if so did that result in losing that code?

**Comment by wasade** [comment]
Didn't squash commits, looks like the stubbed code got dropped [here](https://github.com/wasade/scikit-bio/commit/c7cf6394e5df310252df944e60866d0077a7fe50#diff-8195bb819ac5d477be57b65b6dfb19a4L298)

**Comment by gregcaporaso** [comment]
Thanks!

**Comment by wasade** [comment]
Just syncing out-of-band email discussion about this thread here. A proof of concept block decomposition of the distance matrix is can be found [here](https://github.com/wasade/yummy-octo-duck/blob/master/ipynb/block%20distance%20matrix%20compute.ipynb) with a not-extremely-clean-but-works parallel adaption [here](https://github.com/wasade/yummy-octo-duck/blob/master/ipynb/block%20distance%20matrix%20compute%20parallel.ipynb). This uses a `shear` method that supports operations on tree arrays (i.e., the result of `TreeNode.to_array`) which has the advantage of being about 40x faster when testing against Greengenes, and orders of magnitude less space required for large trees. Tests and discussion on `shear` are [here](https://github.com/wasade/yummy-octo-duck/blob/master/ipynb/Tree%20array%20shearing.ipynb). The `shear` is needed once per block, and as a result, the time and particularly the space requirements are important.

The motivation for the block decomposition method is that it allows for low memory consumption regardless of the number of samples being computed, with the exception of the memory required to represent the resulting distance matrix (which could be mmap'd if needed...). This method shines in particular with methods like unifrac which require an intermediary data structure. In the case of unifrac, that structure scales in space on the order of `O(n * log n * m)` where `n` is the number of OTUs and `m` is the number of samples. By using a smaller `m`, the space is scaled down considerably, with the added benefit that a smaller `m` typically means a smaller `n` too as it is unlikely that all OTUs in the full matrix exist in the subset of samples. The block method effectively pulls out groups of sample IDs at a time, computed beta diversity, and then provides that incremental result which can be filled into the final full matrix. The size of the block is parameterized and is denoted as `k`. A `k` that is too large will result in large amounts of memory due as more samples increases the number of OTUs represented by those samples. However, a small `k` incurs more overhead and more duplicated computation (see below, associated with #1221). 

As a light example, below is a resulting distance matrix with sample IDs `{1, 2, 3, 4, 5, 6}`, and shows what "blocks" are computed using a size of `k=2` where the blocks below are denoted as `{(row_ids), (col_ids)}`:

```text

# final distance matrix
  1 2 3 4 5 6
1 0 a b c d e
2 a 0 f g h i
3 b f 0 j k l
4 c g j 0 m n
5 d h k m 0 o
6 e i l n o 0

# block {(1, 2), (3, 4)}

  1 2 3 4
1 0 a b c
2 a 0 f g
3 b f 0 j
4 c g j 0

# block {(1, 2), (5, 6)}

  1 2 5 6
1 0 a d e
2 a 0 h i
5 d h 0 o
6 e i o 0
 
# block {(3, 4), (5, 6)}

  3 4 5 6
3 0 j k l
4 j 0 m n
5 k m 0 o
6 l n o 0
```

As can be seen above, this approach leads to a large amount of duplicated calculations with `m` is much larger than `k`, on the order of `((m / k) * ((m / k) - 1)` or approximately double the number of calculations when `m=1000` and `k=128`. However, the primary burden of compute is the construction of the intermediary `counts_array` with the subsequent pairwise calculations being very fast. As such, the benefits of low memory and parallelism likely outweigh the expense of additional compute for large tables. There is also potential to mitigate the duplicated compute through modifications associated with #1221.

**Comment by wasade** [comment]
We have highly parallel and optimized solutions in place for UniFrac and Faith's PD maintained external to this library. Parallelization of beta metrics implemented elsewhere, such as Bray Curtis, should depend their respective upstream parallel strategies (i.e., rely on scipy to determine how best to optimize their functions). Parallelization of alpha metrics outside of Faith's PD remains a nice to have.

**Comment by sjanssen2** [comment]
looks like scikit-learn offers parallel computation of beta-diversity metrics: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances.html @qiyunzhu are there any plans to adopt this or do you shy away from making scikit-learn a dependency?

**Comment by qiyunzhu** [comment]
@sjanssen2 Thanks for pointing us to this! Looks cool. The team does have a plan to incorporate parallelization in scikit-bio in a more universal manner. There are several ways to achieve this. scikit-learn uses [joblib](https://joblib.readthedocs.io/en/stable/), which is handy and effective in my tests. Let me forward this note to our software engineer @mataton . Let me also cc @wasade and @mortonjt .
